compiler/rustc_const_eval/src/interpret/intrinsics.rs

```diff
@@ -573,6 +601,106 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
+        assert_matches!(mir_op, BinOp::Add | BinOp::Sub);
+
+        let mir_op = mir_op.wrapping_to_overflowing().unwrap();
+
+        let (val, overflowed1) = self.binary_op(mir_op, l, r)?.to_scalar_pair();
+
+        let val = ImmTy::from_scalar(val, l.layout);
+        let c = ImmTy::from_scalar(c.to_scalar(), l.layout);
+
+        let (val, overflowed2) = self.binary_op(mir_op, &val, &c)?.to_scalar_pair();
+
+        let overflowed1 = overflowed1.to_bool()?;
+        let overflowed2 = overflowed2.to_bool()?;
+
+        let overflowed = Scalar::from_bool(if l.layout.abi.is_signed() {
+            overflowed1 != overflowed2
+        } else {
+            overflowed1 | overflowed2
+        });
+
+        interp_ok((val, overflowed))
+    }
+
+    pub fn mul_double_add2(
+        &self,
+        l: &ImmTy<'tcx, M::Provenance>,
+        r: &ImmTy<'tcx, M::Provenance>,
+        c1: Option<&ImmTy<'tcx, M::Provenance>>,
+        c2: Option<&ImmTy<'tcx, M::Provenance>>,
+    ) -> InterpResult<'tcx, (Scalar<M::Provenance>, Scalar<M::Provenance>)> {
+        assert_eq!(l.layout.ty, r.layout.ty);
+        assert_matches!(l.layout.ty.kind(), ty::Int(..) | ty::Uint(..));
+
+        let is_signed = l.layout.abi.is_signed();
+        let size = l.layout.size;
+        let bits = size.bits();
+        let l = l.to_scalar_int()?;
+        let r = r.to_scalar_int()?;
+
+        interp_ok(if is_signed {
+            let l = l.to_int(size);
+            let r = r.to_int(size);
+            let c1 = c1.map_or(interp_ok(0), |c1| interp_ok(c1.to_scalar_int()?.to_int(size)))?;
+            let c2 = c2.map_or(interp_ok(0), |c2| interp_ok(c2.to_scalar_int()?.to_int(size)))?;
+            if bits == 128 {
+                #[cfg(bootstrap)]
+                {
+                    let _ = (l, r, c1, c2);
+                    unimplemented!()
+                }
+                #[cfg(not(bootstrap))]
+                {
+                    let (lo, hi) = l.carrying2_mul(r, c1, c2);
+                    let lo = Scalar::from_uint(lo, size);
+                    let hi = Scalar::from_int(hi, size);
+                    (lo, hi)
+                }
+            } else {
+                let prod = l * r + c1 + c2;
+                let lo = Scalar::from_int(prod, size);
+                let hi = Scalar::from_int(prod >> size.bits(), size);
+                (lo, hi)
+            }
+        } else {
+            let l = l.to_uint(size);
+            let r = r.to_uint(size);
+            let c1 = c1.map_or(interp_ok(0), |c1| interp_ok(c1.to_scalar_int()?.to_uint(size)))?;
+            let c2 = c2.map_or(interp_ok(0), |c2| interp_ok(c2.to_scalar_int()?.to_uint(size)))?;
+            if bits == 128 {
+                #[cfg(bootstrap)]
+                {
+                    let _ = (l, r, c1, c2);
+                    unimplemented!()
+                }
+                #[cfg(not(bootstrap))]
+                {
+                    let (lo, hi) = l.carrying2_mul(r, c1, c2);
```

Comment:
This is unfortunate... for basic arithmetic like this, it'd be better if we had our own implementation of them rather than having to use the host operation. How hard would that be?

Reply:
So, my thought process here is that the ideal solution is to replace the current hard-coded `i128`/`u128` version of the code with a bigint implementation and do the multiplication directly in all cases. That would be the most ideal, and it would support a potential future with integers larger than 128 bits. It would also likely use methods like this one to perform bigint multiplication.

However, without that, my thought process was that I could either manually code a version of `carrying2_mul` here that would perform worse and require extra scrutiny, or just use the version that's already been implemented and tested.

I'll defer to whatever you think is the better option, but that at least explains my reasoning.

Reply:
I don't know how hard this is to implement directly, is there some code somewhere that would give me an idea?

It's also not great to have a completely different codepath for u128 and the rest, that makes proper testing more tricky.

Reply:
> with a bigint implementation

FWIW I'd love that for all our arithmetic.^^ It's probably too slow though. And using it only sometimes seems odd.

Reply:
> However, without that, my thought process was that I could either manually code a version of carrying2_mul here that would perform worse and require extra scrutiny, or just use the version that's already been implemented and tested.

So, my thought process here is that we typically want to be independent from possible bugs in the standard library, and provide our own reference implementation. But, we haven't done that in numeric_intrinsic, so it'd be odd to use a higher standard here.

So fair, please stick with the current implementation, just with more comments.